courses/advanced-ada/chapters/generics.rst

```diff
@@ -408,3 +392,1511 @@ two:
+
+    end Gen_Float_Acc;
+
+In this case, however, even though we declared a custom addition operator
+for the :ada:`My_Float` type in the :ada:`Float_Types` package, an
+instantiation of the modified :ada:`Gen_Float_Acc` package would always
+make use of the standard addition:
+
+.. code-block:: ada
+
+    --  This makes use of the type definition of My_Float, but not its
+    --  overriden operators.
+    package Float_Ops is new Gen_Float_Acc (F => My_Float);
+
+Because the type :ada:`F` is declared as :ada:`digits <>`, which
+corresponds to the universal floating-point data type, the compiler
+selects operators associated with the universal floating-point data type
+in the package body. By specifying the formal subprogram, we make sure
+that the operator associated with the actual type is used.
+
+Alternatively, we could make use of the :ada:`Float_Types` package
+directly in the generic package. For example:
+
+.. code:: ada
+
+    with Float_Types; use Float_Types;
+
+    generic
+       type F is new My_Float;
+       with function "+" (A : F; B : F) return F is <>;
+    package Gen_Float_Acc is
+       procedure Acc (V : in out F; S : F);
+    end Gen_Float_Acc;
+
+In this case, because the formal type is now based on :ada:`My_Float`, the
+corresponding operator for :ada:`My_Float` is used in the :ada:`Acc`
+procedure.
+
+Generic fixed-point types
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Simple formal package
+^^^^^^^^^^^^^^^^^^^^^
+
+:code-config:`reset_accumulator=True`
+
+In the previous section, we looked into an example of saturation for
+generic floating-point types. Let's adapt this example for fixed-point
+types. This is the package specification:
+
+.. code:: ada
+
+    generic
+       type F is delta <>;
+    package Gen_Fixed_Ops is
+       function Sat_Add (V1, V2 : F) return F;
+    end Gen_Fixed_Ops;
+
+For the fixed-point version, we specify the normalized range in the
+definition of the data type. Therefore, any computation that leads to
+values out of the normalized range will raise a :ada:`Constraint_Error`
+exception. In order to circumvent this, we can declare a fixed-point data
+type with a wider range and use it in combination with the actual
+operation that we want to perform -- an addition, in this case. This
+approach  can be seen in the implementation of :ada:`Sat_Add`, which
+computes the addition using the local :ada:`Ovhd_Fixed` type with wider
+range, calls the :ada:`Saturate` procedure and converts the data type back
+into the original range.
+
+.. code:: ada
+
+    with Ada.Text_IO;    use Ada.Text_IO;
+
+    package body Gen_Fixed_Ops is
+
+       Ovhd_Depth : constant Positive := 64;
+       Ovhd_Bits  : constant := 32;
+       Ovhd_Delta : constant := 2.0 ** Ovhd_Bits / 2.0 ** (Ovhd_Depth - 1);
+
+       type Ovhd_Fixed is delta Ovhd_Delta range
+         -2.0 ** Ovhd_Bits .. 2.0 ** Ovhd_Bits - Ovhd_Delta
+         with Size => Ovhd_Depth;
+
+       --  Ensure that 'First and 'Last have at least double amount
```

Comment:
I don't understand how the assertions that follow ensure what you need. Why not simply

pragma Assert (Ovhd_Fixed'First <= 2 * Ovhd_Fixed(F'First));
pragma Assert (Ovhd_Fixed'Last >= 2 * Ovhd_Fixed(F'Last));

Reply:
Yes, you're right. I've got those from an implementation for a saturated multiplication. For a saturated addition, it's not really needed. I'll change it later.